ansible/roles/influxdb_server/tasks/main.yml

```diff
@@ -0,0 +1,122 @@
+---
+# Copyright (C) 2020 Pedro Luis Lopez <pedroluis.lopezsanchez@gmail.com>
+# Copyright (C) 2020 Innobyte Bechea Leonardo <https://www.innobyte.com/>
+# Copyright (C) 2020 Innobyte Alin Alexandru <https://www.innobyte.com/>
+# Copyright (C) 2020 DebOps <https://debops.org/>
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+- import_role:
+    name: 'ansible_plugins'
+
+- import_role:
+    name: 'secret'
+
+- name: Restart database server on first install (influx bug)
+  service:
+    name: 'influxdb'
+    state: 'restarted'
+  when: not (ansible_local|d() and ansible_local.influxdb_server|d()
+             and ansible_local.influxdb_server.installed|d(false)|bool)
+
+- name: Wait for HTTP endpoint port to become open on the host on first install
+  wait_for:
+    port: '{{ influxdb_server__port }}'
+  when: not (ansible_local|d() and ansible_local.influxdb_server|d()
+             and ansible_local.influxdb_server.installed|d(false)|bool)
+
+- name: Create default admin user on first install
+  influxdb_user:
+    user_name: 'root'
+    user_password: '{{ influxdb_server__root_password }}'
+    state: 'present'
+    admin: 'yes'
+  when: not (ansible_local|d() and ansible_local.influxdb_server|d()
+             and ansible_local.influxdb_server.installed|d(false)|bool)
```

Comment:
I've tried to run this role on a test machine in a tricky environment I manage. In that environment you have to use HTTP proxy to access anything outside local network. So I have:

```yaml
inventory__group_environment:
  http_proxy: 'http://proxy.address.local:8080'
  https_proxy: 'http://proxy.address.local:8080'
```

And at this point the role breaks because i tries to use http proxy to access the server.

Reply:
Afer some testing I found one possible solution:

```yaml
- name: Create default admin user on first install
  influxdb_user:
    user_name: 'root'
    user_password: '{{ influxdb_server__root_password }}'
    state: 'present'
    admin: 'yes'
    proxies: 
      "http": 
      "https": 
  when: not (ansible_local|d() and ansible_local.influxdb_server|d()
             and ansible_local.influxdb_server.installed|d(false)|bool)
```

Reply:
Using `delegate_to` maybe is also required disable proxies in `influxdb` role, tasks in `manage_contents.yml`, right?

An other option is clear http and https proxies environment in the task, like this:

```yaml
- name: Create default admin user on first install
  influxdb_user:
    user_name: 'root'
    user_password: '{{ influxdb_server__root_password }}'
    state: 'present'
    admin: 'yes'
  when: not (ansible_local|d() and ansible_local.influxdb_server|d()
             and ansible_local.influxdb_server.installed|d(false)|bool)```
  environment: 
      http_proxy: ''
      https_proxy: ''
```

But I'm not sure of the environment precedence playboook/task. Can you test it?

Reply:
It works:

```yaml
- name: Create default admin user on first install
  influxdb_user:
    user_name: 'root'
    user_password: '{{ influxdb_server__root_password }}'
    state: 'present'
    admin: 'yes'
  when: not (ansible_local|d() and ansible_local.influxdb_server|d()
             and ansible_local.influxdb_server.installed|d(false)|bool)```
  environment: 
      http_proxy: ''
      https_proxy: ''```

But, with this solution never you will can use proxies for `influxdb_user` and `influxdb_database` tasks.

@drybjed @scibi , what do you think?

Reply:
Also we could disable environment proxies by configuration. Example:

```yaml
influxdb__disable_http_proxies: True
```

```yaml
- name: Create databases
  influxdb_database:
    database_name: '{{ item.database | d(item.name) }}'
    state: 'present'
    hostname: '{{ influxdb__server }}'
    port: '{{ influxdb__port }}'
    ssl: '{{ influxdb__pki }}'
    password: '{{ lookup("password",
                  secret + "/influxdb/" + influxdb__delegate_to +
                  "/credentials/root/password " +
                  "length=" + influxdb__password_length) }}'
  with_flattened: '{{ influxdb__databases + influxdb__dependent_databases|d([]) }}'
  delegate_to: '{{ influxdb__delegate_to }}'
  when: ((item.database|d(False) or item.name|d(False)) and
         (item.state is undefined or item.state != 'absent'))
  environment:
      http_proxy: '{{ "" if influxdb__disable_http_proxies|d(True) else omit }}'
      https_proxy: '{{ "" if influxdb__disable_http_proxies|d(True) else omit }}'
```

Reply:
Both `influxdb_database` and `influxdb_user` support `proxies` option. I thing it's better to use this option than environment variables.

I'm wondering if `influxdb__disable_http_proxies` is really necessary. Why do you use both `delegate_to` and `hostname` when managing users and databases? If you always delegate to machine running InfluxDB you could safely omit `hostname` option so it will always use `localhost`. And in this case you can always disable proxies.

Reply:
Since we are talking REST API here, host FQDN comes into play - the X.509 certificates used for the endpoint will have a specific CN which needs to be the same as the endpoint address we're using, otherwise the TLS connection will be invalid.

I think that delegation should be decoupled from the hostname and by default not enabled at all. Using the `hostname` we can specify the API endpoint address which will be reached from the remote host we are managing. Optionally like in the @scibi's case, we can use `delegate_to` to perform the actual task on a different host, which can actually reach the one specified in the `hostname`. How about this?

Reply:
It's a HTTP API. With SSL it dosen't work with `localhost`, it can't validate certificate names.

For task `Create default admin user on first install` in `influxdb_server` role it isn't a problem, because SSL is not configured at these moment, but it's a problem for tasks in `influxdb` role with a SSL server configured.

Reply:
Anyway, I think I can always disable http proxies. Because it configure users and databases from the same host. But DNS name have to be configured with a IP in the host, vía DNS server or `/etc/hosts`.

Reply:
@drybjed 

So, i can disable http proxies for task `Create default admin user on first install` in `influxdb_server` role.

And drop delegation in `influxdb` role.

Right?

Reply:
Yes, I think that the admin user task doesn't need to be proxied, since we are talking to the server directly anyway over SSH.

The `infludb` role can use `delegate_to` like in the case in the @scibi's environment, but I would set it as `{{ omit }}` by default. This should make it "disabled" in normal usage, and just on special case it can be enabled through the inventory.

Reply:
I've run a test with a current codebase. In all tasks in `manage_contents.yml` I added:
```yaml
proxies:
  http:
  https:
```

And it works.

@drybjed what's the benefit of making `delegate_to` optional?

Reply:
I think that there is another issue with making direct calls to the API from Ansible Controller:

In my case Ansible Controller is my pc. I manage many environments. Sometimes they are created just to test something. All of them use different PKI and I don't add root certificates to trusted CA. When I have to expose some service to users I use external certificates or use Let's Encrypt. I didn't test this but I guess that when `delegate_to` is omitted it will fail to connect because root CA is not trusted.

Reply:
About proxies. For admin task we can disable proxies, it access via localhost. But I don't know if it's a good idea disable proxies and don't allow user to configure proxies if he needs these in `manage_contents.yml` tasks. I think that we can to allow user configure proxies in `influxdb__users` and `influxdb__databases` inventory vars.

About CA's. InfluxDB ansible modules don't allow configure root CA, so it needs a PR to ansible modules. But otherwise, we can allow disable `validate_certs`.

A flexible proposal:

[influxdb_server]
```yaml
- name: Create default admin user on first install
  influxdb_user:
    user_name: 'root'
    user_password: '{{ influxdb_server__root_password }}'
    state: 'present'
    admin: 'yes'
    proxies:
      http:
      https:
  when: not (ansible_local|d() and ansible_local.influxdb_server|d()
             and ansible_local.influxdb_server.installed|d(false)|bool)
```

[influxdb]
```yaml
- name: Create user accounts
  influxdb_user:
    user_name: '{{ item.user | d(item.name) }}'
    user_password: '{{ item.password | d(lookup("password",
                       secret + "/influxdb/" + influxdb__delegate_to +
                       "/credentials/" + item.user | d(item.name) + "/password " +
                       "length=" + influxdb__password_length)) }}'
    grants: '{{ item.grants | d(omit) }}'
    admin: '{{ item.admin | d(False) }}'
    state: 'present'
    hostname: '{{ influxdb__server }}'
    port: '{{ influxdb__port }}'
    ssl: '{{ influxdb__pki }}'
    password: '{{ lookup("password",
                  secret + "/influxdb/" + influxdb__delegate_to +
                  "/credentials/root/password " +
                  "length=" + influxdb__password_length) }}'
    proxies: '{{ item.proxies | d(omit) }}'
    validate_certs: '{{ item.validate_certs | d(True) }}'
  with_flattened: '{{ influxdb__users + influxdb__dependent_users }}'
  delegate_to: '{{ influxdb__delegate_to }}'
  when: ((item.user|d(False) or item.name|d(False)) and
         (item.state is undefined or item.state != "absent"))
```

Reply:
Maybe something like this?

```yaml
proxies: '{{ item.proxies | d(influxdb__proxies | d(omit)) }}'
```

This way it will be easier to write and use roles like `telegraf` which depend on `influxdb`.

Reply:
I have made some changes about `proxies` and `delegate_to`. Can you review it?

Reply:
Hi!

I've set `influxdb__proxies` in my inventory

```yaml
influxdb__proxies:
  http:
  https:
```

And role works great both with and without `influxdb__delegate_to` set.